### PR TITLE
[Feature] Notification systems with new UI panel and helpers to generate messages

### PR DESCRIPTION
**Pull request by macrigiuseppe**
This pull request is a follow up to RFC #307 .

New features:
- Ability to create notifications using helpers
  - __createNotification__
  - __successNotfication__
  - __errorNotfication__
- React components can now dispatch two new actions:
  - __addNotification({message: 'this is a message', type: 'error', topic: 'global, ...})__
  - __removeNotification(notificationId)__
- Two new React components
  - NotificationPanel: container and list of notifications. This component receives the uiState with all notifications and it will only display notifications with the topic set to 'global'
  - NotificationPanelItem: component to display notification message. The new item supports markdown messages.

Both components are injected using dependency injection hence they are also replaceable by using custom components

kepler-notification

### CODE REVIEW

File: src/components/notification-panel/notification-item.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    render() {
+      const {notification, removeNotification} = this.props;
+      return (
+        <NotificationItemContent
+          className={`notification-item`}
+          {...this.props}
+          onClick={() => this.setState({isExpanded: !this.state.isExpanded})}
+          isExpanded={this.state.isExpanded}>
+          <NotificationIcon
+            className={`notification-item--icon`}>
```
[the remark this answers is not in the document]
Comment (macrigiuseppe) [reply src/components/notification-panel/notification-item.js]: yeah, i actually had a variable before in there. I will change it

File: src/components/notification-panel/notification-item.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      this.state = {
+        isExpanded: false
+      };
+    }
+
+    render() {
+      const {notification, removeNotification} = this.props;
+      return (
+        <NotificationItemContent
+          className={`notification-item`}
```
Comment (macrigiuseppe) [inline src/components/notification-panel/notification-item.js]: use regular quotes

File: src/components/notification-panel/notification-item.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    props.theme.notificationPanelItemHeight * (1 + Number(props.isExpanded)) 
+  }px;
+  font-size: 10px;
+  margin-bottom: 1rem;
+  padding: 1em;
+  -webkit-border-radius: 4px;
+  -moz-border-radius: 4px;
+  border-radius: 4px;
+  -webkit-box-shadow: ${props => props.theme.boxShadow};
+  -moz-box-shadow: ${props => props.theme.boxShadow};
```
Comment (CharlieMartell) [inline src/components/notification-panel/notification-item.js]: afaik the styled-components autoprefixer should handle adding these

Comment (macrigiuseppe) [reply src/components/notification-panel/notification-item.js]: thank you

File: src/reducers/ui-state-updaters.js [hunk 4]
```diff
@@ -164,3 +165,16 @@ export const setExportDataUpdater = (state, action) => ({
     data: !state.exportData.data
   }
 });
+
+export const addNotificationUpdater = (state, {payload}) => ({
+  ...state,
+  notifications: [
+    ...state.notifications,
```
Comment (heshan0131) [inline src/reducers/ui-state-updaters.js]: `...(state.notifications || [])`

File: src/components/kepler-gl.js [hunk 5]
```diff
@@ -279,6 +287,9 @@ function KeplerGlFactory(
               this.root = node;
             }}
           >
+            {uiState.notifications.length !== 0 ?
```
Comment (heshan0131) [inline src/components/kepler-gl.js]: even if `uiState.notifications.length === 0` `NotificationPanel` should still work right? I would profer apply less logic here and just pass `uiState.notifications` to `< NotificationPanel>` to let it figure out what to do.

File: src/components/kepler-gl.js [hunk 6]
```diff
@@ -204,6 +207,11 @@ function KeplerGlFactory(
         clicked
       } = visState;
 
+      const notificationPanelFields = {
+        uiStateActions,
+        uiState
```
Comment (heshan0131) [inline src/components/kepler-gl.js]: If you are not using other properties in `uiState`, should just pass in `uiState.notifications` here, making the `notifications -> notificationPanel` logic simple and clear

File: src/reducers/ui-state-updaters.js (hunk 4 above)
Comment (macrigiuseppe) [reply src/reducers/ui-state-updaters.js]: notifications has a default value in the reducer. Do you still want to apply the OR statement?

File: src/components/kepler-gl.js (hunk 6 above)
Comment (macrigiuseppe) [reply src/components/kepler-gl.js]: I have been thinking about this, in fact the first version had only notifications. I thought that passing the full uiState would provide more information and context about the current state of the UI. Currently we only use notifications but maybe in the future we could take advantage of other properties to change the notification panel behavior

File: examples/demo-app/src/app.js [hunk 7]
```diff
@@ -147,6 +150,26 @@ class App extends Component {
     window.localStorage.setItem(BannerKey, 'true');
   };
 
+  /* eslint-disable max-nested-callbacks */
+  _loadMockNotifications = () => {
+    setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (macrigiuseppe) [reply examples/demo-app/src/app.js]: BTW this is just for showing an example of notifications. I will do it, i really like the approach

File: src/components/kepler-gl.js (hunk 5 above)
Comment (macrigiuseppe) [reply src/components/kepler-gl.js]: done

File: src/components/kepler-gl.js (hunk 6 above)
Comment (macrigiuseppe) [reply src/components/kepler-gl.js]: Let's only use notificaitons and removeNotification as action

**Pull request merged**